sync/class.jetpack-sync-module-posts.php

```diff
@@ -55,75 +59,42 @@ public function init_listeners( $callable ) {
 
 		// WordPress, Blogger, Livejournal, woo tax rate
 		add_action( 'import_end', array( $this, 'sync_import_end' ) );
+
+		add_action( 'set_object_terms', array( $this, 'set_object_terms' ), 10, 6 );
 	}
 
-	public function sync_import_done( $importer ) {
-		// We already ran an send the import
-		if ( $this->import_end ) {
-			return;
+	public function set_post_sync_item( $post_parent, $post_ID ) {
+		if ( $post_ID ) {
+			$this->sync_items[ $post_ID ] = new Jetpack_Sync_Item( 'save_post' );
+		} else {
+			//
+			$this->sync_items[ 'new' ] = new Jetpack_Sync_Item( 'save_post' );
 		}
-
-		$importer_name = $this->get_importer_name( $importer );
-
-		/**
-		 * Sync Event that tells that the import is finished
-		 *
-		 * @since 5.0.0
-		 *
-		 * $param string $importer
-		 */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
-		$this->import_end = true;
+		return $post_parent;
 	}
 
-	public function sync_import_end() {
-		// We already ran an send the import
-		if ( $this->import_end ) {
+	public function set_object_terms( $post_id, $terms, $tt_ids, $taxonomy, $append, $old_tt_ids ) {
+		if ( ! self::is_saving_post( $post_id ) ) {
 			return;
 		}
-
-		$this->import_end = true;
-		$importer         = 'unknown';
-		$backtrace        = wp_debug_backtrace_summary( null, 0, false );
-		if ( $this->is_importer( $backtrace, 'Blogger_Importer' ) ) {
-			$importer = 'blogger';
-		}
-
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WC_Tax_Rate_Importer' ) ) {
-			$importer = 'woo-tax-rate';
-		}
-
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WP_Import' ) ) {
-			$importer = 'wordpress';
-		}
-
-		$importer_name = $this->get_importer_name( $importer );
-
-		/** This filter is already documented in sync/class.jetpack-sync-module-posts.php */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
-	}
-
-	private function get_importer_name( $importer ) {
-		$importers = get_importers();
-		return isset( $importers[ $importer ] ) ? $importers[ $importer ][0] : 'Unknown Importer';
+		$sync_item = new Jetpack_Sync_Item( 'set_object_terms',
+			array( $post_id, $terms, $tt_ids, $taxonomy, $append, $old_tt_ids )
+		);
+		$this->sync_items[ $post_id ]->add_sync_item( $sync_item );
 	}
 
-	private function is_importer( $backtrace, $class_name ) {
-		foreach ( $backtrace as $trace ) {
-			if ( strpos( $trace, $class_name ) !== false ) {
-				return true;
-			}
-		}
 
-		return false;
+	public function is_saving_post( $post_ID ) {
+		return isset( $this->sync_items[ $post_ID ] );
 	}
 
 	public function init_full_sync_listeners( $callable ) {
 		add_action( 'jetpack_full_sync_posts', $callable ); // also sends post meta
 	}
 
 	public function init_before_send() {
```

Comment:
In the next phase of refactoring, I'd love to see the sender call a `process_before_send()` method on each sync item, where each sync item is a subclass containing the code necessary to do whatever processing it needs to, rather than this mechanism.